core/src/main/scala/kafka/zk/AdminZkClient.scala

```diff
@@ -0,0 +1,417 @@
+    createOrUpdateTopicPartitionAssignmentPathInZK(topic, replicaAssignment, topicConfig)
+  }
+
+  /**
+   * Gets broker metadata list
+   * @param rackAwareMode
+   * @param brokerList
+   * @return
+   */
+  def getBrokerMetadatas(rackAwareMode: RackAwareMode = RackAwareMode.Enforced,
+                         brokerList: Option[Seq[Int]] = None): Seq[BrokerMetadata] = {
+    val allBrokers = zkClient.getAllBrokersInCluster
+    val brokers = brokerList.map(brokerIds => allBrokers.filter(b => brokerIds.contains(b.id))).getOrElse(allBrokers)
+    val brokersWithRack = brokers.filter(_.rack.nonEmpty)
+    if (rackAwareMode == RackAwareMode.Enforced && brokersWithRack.nonEmpty && brokersWithRack.size < brokers.size) {
+      throw new AdminOperationException("Not all brokers have rack information. Add --disable-rack-aware in command line" +
+        " to make replica assignment without rack information.")
+    }
+    val brokerMetadatas = rackAwareMode match {
+      case RackAwareMode.Disabled => brokers.map(broker => BrokerMetadata(broker.id, None))
+      case RackAwareMode.Safe if brokersWithRack.size < brokers.size =>
+        brokers.map(broker => BrokerMetadata(broker.id, None))
+      case _ => brokers.map(broker => BrokerMetadata(broker.id, broker.rack))
+    }
+    brokerMetadatas.sortBy(_.id)
+  }
+
+  /**
+   * Creates or Updates the partition assignment for a given topic
+   * @param topic
+   * @param partitionReplicaAssignment
+   * @param config
+   * @param update
+   */
+  def createOrUpdateTopicPartitionAssignmentPathInZK(topic: String,
+                                                     partitionReplicaAssignment: Map[Int, Seq[Int]],
+                                                     config: Properties = new Properties,
+                                                     update: Boolean = false) {
+    validateCreateOrUpdateTopic(topic, partitionReplicaAssignment, config, update)
+
+    // Configs only matter if a topic is being created. Changing configs via AlterTopic is not supported
+    if (!update) {
+      // write out the config if there is any, this isn't transactional with the partition assignments
+      zkClient.setOrCreateEntityConfigs(ConfigType.Topic, topic, config)
+    }
+
+    // create the partition assignment
+    writeTopicPartitionAssignment(topic, partitionReplicaAssignment, update)
+  }
+
+  /**
+   * Validate method to use before the topic creation or update
+   * @param topic
+   * @param partitionReplicaAssignment
+   * @param config
+   * @param update
+   */
+  def validateCreateOrUpdateTopic(topic: String,
+                                  partitionReplicaAssignment: Map[Int, Seq[Int]],
+                                  config: Properties,
+                                  update: Boolean): Unit = {
+    // validate arguments
+    Topic.validate(topic)
+
+    if (!update) {
+      if (zkClient.topicExists(topic))
+        throw new TopicExistsException(s"Topic '$topic' already exists.")
```

Comment:
Do we need this check here since writeTopicPartitionAssignment() already covers it?

Reply:
there are few direct callers for validateCreateOrUpdateTopic method. so we need the check